mac: update permission requirements for DD v.4.15 (#16247)
docker/docs: desktop/mac/permission-requirements.md

- /desktop/mac/privileged-helper/
---

This page contains information about the permission requirements for running and installing Docker Desktop on Mac.

It also provides clarity on running containers as `root` as opposed to having `root` access on the host.

## Permission requirements

> Docker Desktop for Mac is run as an unprivileged user. However, certain funcitonalities may require for Docker Desktop to perform a limited set of privileged configurations such as:

Reviewer: ```suggestion
Docker Desktop for Mac is run as an unprivileged user. However, certain functionalities may be required for Docker Desktop to perform a limited set of privileged configurations such as:
```

> Docker Desktop for Mac is run as an unprivileged user. However, certain functionalities may be required for Docker Desktop to perform a limited set of privileged configurations such as:

Reviewer 2: `However, certain functionalities may be required for Docker Desktop to perform a limited set of privileged configurations` 
doesn't sound right. Perhaps:
`However, certain functionalities may require Docker Desktop to perform a limited set of privileged configurations` ?

>  - Installing symlinks in `/usr/local/bin`. This ensures the `docker` CLI is on the user’s PATH without having to reconfigure shells, log out then log back in for example.

Reviewer: ```suggestion
 - Installing symlinks in `/usr/local/bin`. This ensures the `docker` CLI is on the user’s PATH without having to reconfigure shells, log out then log back in, for example.
```

 - Securely caching the Registry Access Management policy which is read-only for the developer.

Versions prior to Docker Desktop v4.15, do require `root` access to be granted on the first run. The first time that Docker Desktop is launched the user receives an admin prompt to grant permissions for a privileged helper service `com.docker.vmnetd` to be installed. For subsequent runs, no `root` privileges are required. This approach allowed, following the principle of least privilege, `root` access to be used only for the operations for which it is absolutely necessary, while still being able to use Docker Desktop as an unprivileged user. 
All privileged operations are run using the privileged helper process `com.docker.vmnetd`.

For security reasons, starting with +v4.15, Docker Desktop for Mac does not require to run a permanent privileged process. Whenever elevated privileges are needed for a configuration,
Docker Desktop prompts the user with information on the task it needs to perform. Most configurations are applied once, subsequent runs do not prompt for privileged access anymore.

> The only time, Docker Desktop may start the privileged process is for binding privileged ports that are not allowed by default on the host OS.

Reviewer: ```suggestion
The only time Docker Desktop may start the privileged process is for binding privileged ports that are not allowed by default on the host OS.
```

- creates the symlinks without the admin prompt if `/usr/local/bin` is writable by unprivileged users
- triggers an admin prompt for user to authorize the creation of symlinks in `/usr/local/bin`. If user authorizes, symlinks to docker binaries are created in `/usr/local/bin`.
If user rejects the prompt, not willing to run configurations requiring elevated privileges or due to not having admin rights on its machine, Docker Desktop falls back to create the symlinks in `~/.docker/bin` and edit user's shell profile to ensure this location is in user's PATH. This requires all open shells to be reloaded. 
The rejection is recorded for future runs to avoid prompting the user again.
For any failure to ensure binaries are on user's PATH, the user may need to add manually to its PATH the `/Applications/Docker.app/Contents/Resources/bin` or use full path to docker binaries.

A particular case is the installation of the `/var/run/docker.sock` symlink. Creating this symlink ensures various docker clients relying on the default docker socket path to work without additional changes. As the `/var/run` is mounted as a tmpfs, its content is deleted on restart, symlink to docker socket included.

> To ensure the docker socket exists after restart, Docker Desktop sets up a launchd startup task that creates the symlink by running `ln -s -f /Users/<user>/.docker/run/docker.sock /var/run/docker.sock`. This ensures, the user is not prompted all the time on startup to create the symlink. If user rejects the prompt, the symlink and the startup task is not created and user may have to explicitly set the DOCKER_HOST to `/Users/<user>/.docker/run/docker.sock` in the clients it is using. The docker CLI relies on the current context to retrieve the socket path, current context is set to `desktop-linux` on Docker Desktop startup.

Reviewer: ```suggestion
To ensure the docker socket exists after restart, Docker Desktop sets up a launched startup task that creates the symlink by running `ln -s -f /Users/<user>/.docker/run/docker.sock /var/run/docker.sock`. This ensures the user is not prompted all the time on startup to create the symlink. If the user rejects the prompt, the symlink and the startup task are not created and the user may have to explicitly set the DOCKER_HOST to `/Users/<user>/.docker/run/docker.sock` in the clients it is using. The docker CLI relies on the current context to retrieve the socket path, the current context is set to `desktop-linux` on Docker Desktop startup.
```

### Binding privileged ports

> When running container that require privileged ports binding, Docker Desktop first attempts to bind it directly as unprivileged process. In case of failure due to the OS preventing it, Docker Desktop checks if the `com.docker.vmnetd` privileged helper process is running to bind the privileged port through it. If the privileged helper process is not running, Docker Desktop prompts the user for authorization to run it under [launchd](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html).

Reviewer: ```suggestion
When running a container that requires privileged ports binding, Docker Desktop first attempts to bind it directly as an unprivileged process. In case of failure due to the OS preventing it, Docker Desktop checks if the `com.docker.vmnetd` privileged helper process is running to bind the privileged port through it. If the privileged helper process is not running, Docker Desktop prompts the user for authorization to run it under [launchd](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html).
```

> This configures the privileged helper to run as in the versions of Docker Desktop prior to v4.15. However, the functionality provided by this privileged helper has been stripped down. It only supports now port binding and caching the Registry Access Management policy.

Reviewer: ```suggestion
This configures the privileged helper to run as in the versions of Docker Desktop prior to v4.15. However, the functionality provided by this privileged helper has been stripped down. It now only supports port binding and caching the Registry Access Management policy.
```

> If user declines the launch of the privileged helper process, the privileged port binding cannot be done and the docker CLI returns an error:

Reviewer: ```suggestion
If the user declines the launch of the privileged helper process, the privileged port binding cannot be done and the docker CLI returns an error:
```

The removal of the privileged helper process can be done in the typical way of removing launchd processes.

```console
$ ps aux | grep vmnetd                                                
root             28739   0.0  0.0 34859128    228   ??  Ss    6:03PM   0:00.06 /Library/PrivilegedHelperTools/com.docker.vmnetd
user             32222   0.0  0.0 34122828    808 s000  R+   12:55PM   0:00.00 grep vmnetd
$ sudo launchctl unload -w /Library/LaunchDaemons/com.docker.vmnetd.plist
Password:
$ ps aux | grep vmnetd                                                   
user             32242   0.0  0.0 34122828    716 s000  R+   12:55PM   0:00.00 grep vmnetd
$ rm /Library/LaunchDaemons/com.docker.vmnetd.plist 
$ rm /Library/PrivilegedHelperTools/com.docker.vmnetd 
```

### Ensuring `localhost` and `kubernetes.docker.internal` are defined

On first run, Docker Desktop checks if `localhost` is resolved to `127.0.0.1`. In case the resolution fails, it prompts the user to allow adding the mapping to `/etc/hosts`. Similarly, when the Kubernetes cluster is installed, it checks that `kubernetes.docker.internal` is resolved to `127.0.0.1` and prompts the user to do so. 

> ## Privileged Helper

Reviewer 2: I suggest to remove the "Privileged helper" section fully, as it is confusing now that we said that _the functionality provided by this privileged helper has been stripped down. It now only supports port binding and caching the Registry Access Management policy._

Or, we can keep it, but change it to something like this (although it feels like a repetition of what we already said above):

```
In the limited situations when the privileged helper is needed (i.e. binding privileged ports or caching the Registry Access Management policy), the privileged helper is started by `launchd` and runs in the background unless it is disabled at runtime as previously described. The Docker Desktop backend communicates with it over the UNIX domain socket `/var/run/com.docker.vmnetd.sock`. The functionalities it performs are: 
- Binding privileged ports that are less than 1024.
- Securely caching the Registry Access Management policy which is read-only for the developer.
- Uninstalling the privileged helper.
```

The privileged helper is started by `launchd` and runs in the background unless it is disabled at runtime as previously described. The Docker Desktop backend communicates with it over the UNIX domain socket `/var/run/com.docker.vmnetd.sock`. The functionalities it performs are: 
- Installing and uninstalling symlinks in `/usr/local/bin`. 
- Binding privileged ports that are less than 1024.
- Ensuring `localhost` and `kubernetes.docker.internal` are defined in `/etc/hosts`. 
- Securely caching the Registry Access Management policy which is read-only for the developer.
- Performing some diagnostic actions, in particular gathering a performance trace of Docker itself.
- Uninstalling the privileged helper.

> In version 4.11 and later of Docker Desktop for Mac you can avoid running the privileged helper service in the background by using the `--user` flag on the [install command](../install/mac-install.md#install-from-the-command-line). This results in `com.docker.vmnetd` being used for set up during installation and then disabled at runtime. In this case, the user is  not prompted to grant root privileges on the first run of Docker Desktop. Specifically, the `--user` flag:

Reviewer 2: This is not accurate anymore, as we are no longer running the privileged helper service in the background. Using the --user flag performs the privileged configurations at install time, so that the user does not get the prompt on startup. 
So this paragraph does not belong to the "Privileged helper" section, but perhaps in its own section, something like "Performing privileged configurations at installation time".

It also provides clarity on running containers as `root` as opposed to having `root` access on the host.

## Permission requirements

Docker Desktop for Mac is run as an unprivileged user. However, certain functionalities may require for Docker Desktop to perform a limited set of privileged configurations such as:
 - Installing symlinks in `/usr/local/bin`. This ensures the `docker` CLI is on the user’s PATH without having to reconfigure shells, log out then log back in, for example.
 - Binding privileged ports that are less than 1024. The so-called "privileged ports" have not generally been used as a security boundary, however OSes still prevent unprivileged processes from binding them which breaks commands like `docker run -p 127.0.0.1:80:80 docker/getting-started`
 - Ensuring `localhost` and `kubernetes.docker.internal` are defined in `/etc/hosts`. Some old macOS installs did not have `localhost` in `/etc/hosts`, which caused Docker to fail. Defining the DNS name `kubernetes.docker.internal` allows us to share Kubernetes contexts with containers.
 - Securely caching the Registry Access Management policy which is read-only for the developer.

> Versions prior to Docker Desktop v4.15, do require `root` access to be granted on the first run. The first time that Docker Desktop is launched the user receives an admin prompt to grant permissions for a privileged helper service `com.docker.vmnetd` to be installed. For subsequent runs, no `root` privileges are required. This approach allowed, following the principle of least privilege, `root` access to be used only for the operations for which it is absolutely necessary, while still being able to use Docker Desktop as an unprivileged user. 

Reviewer 2: ```suggestion
Versions prior to Docker Desktop 4.15, require `root` access to be granted on the first run. The first time that Docker Desktop is launched the user receives an admin prompt to grant permissions for a privileged helper service `com.docker.vmnetd` to be installed. For subsequent runs, no `root` privileges are required. This approach allowed, following the principle of least privilege, `root` access to be used only for the operations for which it is absolutely necessary, while still being able to use Docker Desktop as an unprivileged user. 
```

All privileged operations are run using the privileged helper process `com.docker.vmnetd`.

> For security reasons, starting with +v4.15, Docker Desktop for Mac does not require the user to run a permanent privileged process. Whenever elevated privileges are needed for a configuration,

Reviewer 2: ```suggestion
For security reasons, starting with the version 4.15, Docker Desktop for Mac does not require the user to run a permanent privileged process. Whenever elevated privileges are needed for a configuration,
```

Docker Desktop prompts the user with information on the task it needs to perform. Most configurations are applied once, subsequent runs do not prompt for privileged access anymore.
The only time Docker Desktop may start the privileged process is for binding privileged ports that are not allowed by default on the host OS.

### Installing symlinks

The docker binaries are installed by default in `/Applications/Docker.app/Contents/Resources/bin`. Docker Desktop ensures the `docker` CLI is on the user’s PATH without having to reconfigure shells, log out then log back in for example. As on most systems, `/usr/local/bin` is by default in user's PATH, Docker Desktop creates symlinks for all docker binaries in it.

> Installing symlinks in `/usr/local/bin` is a privileged configuration Docker Desktop performed on the first startup. Docker Desktop checks if symlinks exists and takes the following actions: 

Reviewer 2: ```suggestion
Installing symlinks in `/usr/local/bin` is a privileged configuration Docker Desktop performs on the first startup. Docker Desktop checks if symlinks exists and takes the following actions: 
```

### Binding privileged ports

When running a container that requires privileged ports binding, Docker Desktop first attempts to bind it directly as an unprivileged process. In case of failure due to the OS preventing it, Docker Desktop checks if the `com.docker.vmnetd` privileged helper process is running to bind the privileged port through it. If the privileged helper process is not running, Docker Desktop prompts the user for authorization to run it under [launchd](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html).

> This configures the privileged helper to run as in the versions of Docker Desktop prior to v4.15. However, the functionality provided by this privileged helper has been stripped down. It now only supports port binding and caching the Registry Access Management policy.

Reviewer 2: ```suggestion
This configures the privileged helper to run as in the versions of Docker Desktop prior to 4.15. However, the functionality provided by this privileged helper has been stripped down. It now only supports port binding and caching the Registry Access Management policy.
```

If the user declines the launch of the privileged helper process, the privileged port binding cannot be done and the docker CLI returns an error:
```console
$ docker run -p 127.0.0.1:80:80 docker/getting-started

docker: Error response from daemon: Ports are not available: exposing port
TCP 127.0.0.1:80 -> 0.0.0.0:0: failed to connect to /var/run/com.docker.vmnetd.sock:
is vmnetd running?: dial unix /var/run/com.docker.vmnetd.sock: connect: connection
refused.
ERRO[0003] error waiting for container: context canceled
```

> **Note**
>
> The command may fail with the same error if user takes too long to authorize the prompt to start the helper process as it may timeout.

> The removal of the privileged helper process can be done in the typical way of removing launchd processes.

Reviewer 2: Shall we move these instructions about removal to the `Privileged Helper` section?

> - Binding privileged ports (<1024) on `127.0.0.1` will not work. For example, `docker run -p 127.0.0.1:80:80 docker/getting-started` will fail, `docker run -p 80:80 docker/getting-started` however will succeed as binding privileged ports on `0.0.0.0` is no longer a privileged operation on recent versions of MacOS.

Reviewer 2: remove this bullet point, as binding privileged ports can still be done if the user authorises the privileged service. Besides, this is already covered in the "Binding privileged ports" section
